test/studies/ml/lib/Chai.chpl

```diff
@@ -0,0 +1,828 @@
+                const dZ_dW: Tensor(1) = flattened;
+                const dZ_dB: real = 1;
+                const dZ_dIn: Tensor(2) = weights;
+                const dL_dZ: Tensor(1) = dL_dOut[i] * dOut_dZ;
+                const dL_dW: Tensor(2) = dL_dZ * dZ_dW.transpose();
+                const dL_dB: Tensor(1) = dL_dZ * dZ_dB;
+                const dL_dIn: Tensor(1) = dZ_dIn.transpose() * dL_dZ;
+                weightsGrad += dL_dW;
+                biasesGrad += dL_dB;
+                newDeltas[idx] = dL_dIn.reshape(input.domain);;
+            }
+            this.weightsGrad.data = weightsGrad.data;
+            this.biasesGrad.data = biasesGrad.data;
+            return newDeltas;
+        }
+
+        proc optimize(mag: real(64)) {
+            weights.data -= mag * weightsGrad.data;
+            biases.data -= mag * biasesGrad.data;
+        }
+        proc resetGradients() {
+            weightsGrad.data = 0.0;
+            biasesGrad.data = 0.0;
+        }
+
+        proc write(fw: IO.fileWriter) throws {
+            weights.write(fw);
+            biases.write(fw);
+        }
+        proc read(fr: IO.fileReader) throws {
+            weights.read(fr);
+            biases.read(fr);
+            uninitialized = false;
+        }
+
+        proc signature(): string {
+            return "SoftMax(" + outputSize:string + ")";
+        }
+
+    }
+
+
+    proc forwardPropHelp(ref layers, param n: int, x: Tensor(?)) {
+        if n == layers.size then return x;
+
+        const xNext = layers[n].forwardProp(x);
+        return forwardPropHelp(layers, n+1, xNext);
+    }
+
+
+
+    proc backwardPropHelp(ref layers, param n: int, x: Tensor(?)) {
+        if n == 0 then return layers[0].backward(x);
+
+        const xNext = layers[n].backward(x);
+        return backwardPropHelp(layers, n-1, xNext);
+    }
+
+    proc backwardForwardPropHelp(ref layers, param n: int, x: Tensor(?), lastDelta: Tensor(?)) {
+
+        if n == layers.size - 1 {
+            return layers[n].backward(lastDelta,x);
+        }
+
+        const lastInput = layers[n].forwardProp(x);
+        const delta = backwardForwardPropHelp(layers, n+1, lastInput, lastDelta);
+        return layers[n].backward(delta,x);
+    }
+
+    proc forwardPropHelpBatch(ref layers, param n: int, inputs) {
+        if n == layers.size then return inputs;
+        const nextInputs = layers[n].forwardPropBatch(inputs);
+        return forwardPropHelpBatch(layers, n+1, nextInputs);
+    }
+
+    proc backwardForwardPropHelpBatch(ref layers, param n: int, inputs, lastDeltas) {
+        if n == layers.size { return lastDeltas; }
+
+        const lastInputs = layers[n].forwardPropBatch(inputs);
+        const deltas = backwardForwardPropHelpBatch(layers, n+1, lastInputs, lastDeltas);
+        return layers[n].backwardBatch(deltas=deltas,inputs=inputs);
+    }
+
+    record Network {
+        var _layers;
+        proc ref layers ref do return this._layers;
+
+        proc init(layers) {
```

Comment:
Are we able to constrain the type of `layers` at all? I think we might have discussed this at some point, but I don't remember if something like `init(layers: ?t) where isTupleType(t)` worked.

Reply:
I don't think so, since each layer type has no relation to each other except that they just so happen to share the same needed procs. This could be constrained using an interface, however. I just wasn't sure when interfaces would be available.

Reply:
Actually, I am not certain this would work for interfaces. Since the tensor rank changes depending on the layer, it would have to be possible to write ```chapel
interface Layer {
    proc forwardProp(batch: [] Tensor): [] Tensor;
}
```

Reply:
Hmm, yeah, or maybe more explicit genericism is (or will be) allowed:
```chapel
proc forwardProp(batch: [] Tensor(?)): [] Tensor(?)
}
```

But I was thinking it would be nice to at least assert that the argument is a tuple. Does that where-clause I wrote above work?

Reply:
Yes it does. I added it. I also added another initializer so that you don't need to wrap the layer sequence in a tuple. The changes are present in the latest commit.